modules/docs/src/02-the-smithy-idl/01-smithy-idl.md

```diff
@@ -0,0 +1,199 @@
+
+### Primitive types
+
+Smithy provides the following "primitive" types out of the box.
+
+* Boolean
+* String
+* Integer
+* Long
+* Float
+* Short
+* Double
+* Byte
+* BigInteger
+* BigDecimal
+* Blob (`schematic.ByteArray`, wrapper to `Array[Byte]`)
+* Timestamp (`smithy4s.Timestamp`, translated from/to java or javascript time types)
+* Document (`smithy4s.Document`, a bespoke Json ADT)
+
+### Named primitives
+
+Smithy lets you define custom names for primitive types :
+
+```kotlin
+namespace foo
+
+integer Age
+long Identifier
+```
+
+These get translated as unboxed type wrappers, or `newtypes`, that look like a case class but do not induce any boxing at runtiime.
+
+### Collection types
+
+Smithy provides 3 different shapes of collections : lists, sets, and maps. They translate to the same, in the generated Scala code.
+
+```kotlin
+namespace foo
+
+list IntList {
+  member: Integer
+}
+
+set StringSet {
+  member: Set
+}
+
+// At this time, only string shapes can be used as keys to map.
+map AgeMap {
+  key: String,
+  value: Integer
+}
+```
+
+### Structures
+
+Structures are product types. In scala, they naturally translate to case classes in Scala.
+
+```kotlin
+namespace foo
+
+structure Person {
+  @required
+  firstName: String,
+  @required
+  lastName: String,
+  dateOfBirth: Timestamp
+}
+```
+
+### Unions
+
+Unions are coproduct types. In Scala, they quite naturally translate to sealed-traits.
+
+Union members can target any data shape, be that a structure or a primitive type.
+
+```kotlin
+namespace foo
+
+structure Cat {
+  name: String
+}
+
+structure Dog {
+  name: String
+}
+
+union Animal {
+  cat: Cat,
+  dog: Dog
+}
+```
+
+### Operations and services
+
+#### Operations
+
+Operations are essentially an optional Input, an optional Output, and a optional list of errors. Inputs, outputs and errors all have to be structure shapes.
+
+```kotlin
+namespace foo
+
+operation Greet {
+  input: GreetInput,
+  output: GreetOutput,
+  errors: [BadInput]
+}
+
+structure GreetInput {
+  name: String
+}
+
+structure GreetOutput {
+  message: String
+}
+
+@error("client")
+structure BadInput {
+  message: String
+}
+```
+
+#### Errors
+
+Regarding errors, `smithy4s` translates them as case classes extending `Throwable`. If the structure shape contains a `message` field, the `getMessage` method of the throwable is implemented in terms of this field.
+
+
+#### Services
+
+Services are basically a list of operations, and an optional list of errors.
+
+```kotlin
+namespace foo
+
+service HelloService {
+  operations: [Greet],
+  errors: [ServerError]
+}
+
+@error("server")
+structure ServerError {
+  message: String
+}
+```
+
+Smithy4s translates them in the following fashion :
+
+```scala
+package foo
+
+trait HelloServiceGen[P[_, _, _, _, _]]{
```

Comment:
I think we should be showing the monadic variant as the first one, since it'll be the major usecase (the only one at the moment, really)